---
title: "L0 -- Course Overview"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{L0 -- Course Overview}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

```{r setup, echo=FALSE, messages=FALSE, warnings=FALSE}
```

Author: Martin Morgan
  (<a href="mailto:mtmorgan@fhcrc.org">mtmorgan@fhcrc.org</a>)<br />
Date: 26-28 October, 2015<br />

The material in this course requires _R_ version 3.2 and
_Bioconductor_ version 3.2

```{r configure-test}
stopifnot(
    getRversion() >= '3.2' && getRversion() < '3.3',
    BiocInstaller::biocVersion() == "3.2"
)
```

This course provides a broad introduction to the analysis and
comprehension of high-throughput sequence data using R and
_Bioconductor_; many other useful software tools are not discussed.  The
topics and learning objectives to be address are as follows:

- [Introduction to _Bioconductor_](L1-Introduction.html): Overall
  sequence analysis work flows. Familiarity with _R_. The diversity of
  analysis challenges addressed by _Bioconductor_. The role of
  _Bioconductor_ packages in specific work flow steps. _Bioconductor_
  resources for help and effective use. Primary packages:
  [GenomicRanges][], [SummarizedExperiment][].

- [Sequence Data Representation and Manipulation](L2-Data-Representation.html):
  Common genomic file types (FASTA/FASTQ, BED, WIG, GTF, etc) and
  their _Bioconductor_ representation and manipulation. Essential
  _Bioconductor_ data structures (e.g., DNAStringSet,
  GenomicRanges). _Bioconductor_ gene and genome 'annotation'
  resources. Simple _Bioconductor_ solutions to common bioinformatic
  tasks. Primary packages: [AnnotationHub][], [rtracklayer][],
  [GenomicFeatures][], [GenomicAlignments][], [ShortRead][],
  [Biostrings][].

- [Variant Calling and Annotation](L3-Variants.html): Conceptual
  understanding of variant calling approaches and (primarily
  non-_Bioconductor_) tools. Exploratory approaches to variant calling
  in _Bioconductor_. Working with called variants, e.g., classifying
  variants to genomic regions of interest and predicting effects of
  variants. Primary packages: [VariantTools][], [r5vc][],
  [VariantAnnotation][], [VariantFiltering][].

- [RNA-Seq Differential Expression](L4-RNASeq.html): _Bioconductor_
  known-gene RNA-seq differential expression work flow, from aligned
  reads to differential expression of genes. Important statistical
  issues and their resolution. Placing results of differential
  expression analysis into biological context. Brief discussion of
  novel-gene and transcript-level RNAseq differential expression
  analysis. Primary packages: [DESeq2][], [edgeR][].

- [Working with Large Data](L5-Large-Data.html): Overcoming common
  pitfalls in memory management and computational throughput when
  writing R code. Adopting restriction, iteration, and parallel
  evaluation strategies for working with large data. Primary packages:
  [BiocParallel][], [GenomicFiles][].

- [Data Visualization](L6-Visualization.html): Challenges to
  visualizing genomic data. Opportunities for familiar and novel
  genomic visualizations in _Bioconductor_. Approaches to interactive
  visualization with shiny. Adopting strategies for reproducible and
  collaborative research. Primary packages: [Gviz][],
  [ReportingTools][], [shiny][], ([ggbio][]).

We'll cover this material using several different approaches:

- _Presentation_ in more-or-less traditional lectures
- _Interactivities_ where we work through material together
- _Lab exercises_ where you work alone or in small groups to complete
  a task.

In addition, I hope that we can have short (5-10 minute) presentations
from course participants about their own research.

[AnnotationHub]: http://bioconductor.org/packages/AnnotationHub
[BiocParallel]: http://bioconductor.org/packages/BiocParallel
[Biostrings]: http://bioconductor.org/packages/Biostrings
[DESeq2]: http://bioconductor.org/packages/DESeq2
[GenomicAlignments]: http://bioconductor.org/packages/GenomicAlignments
[GenomicFeatures]: http://bioconductor.org/packages/GenomicFeatures
[GenomicFiles]: http://bioconductor.org/packages/GenomicFiles
[GenomicRanges]: http://bioconductor.org/packages/GenomicRanges
[Gviz]: http://bioconductor.org/packages/Gviz
[ReportingTools]: http://bioconductor.org/packages/ReportingTools
[ShortRead]: http://bioconductor.org/packages/ShortRead
[SummarizedExperiment]: http://bioconductor.org/packages/SummarizedExperiment
[VariantAnnotation]: http://bioconductor.org/packages/VariantAnnotation
[VariantFiltering]: http://bioconductor.org/packages/VariantFiltering
[VariantTools]: http://bioconductor.org/packages/VariantTools
[edgeR]: http://bioconductor.org/packages/edgeR
[ggbio]: http://bioconductor.org/packages/ggbio
[r5vc]: http://bioconductor.org/packages/r5vc
[rtracklayer]: http://bioconductor.org/packages/rtracklayer
[shiny]: http://shiny.rstudio.org
