---
title: "L3 -- Variant Calling and Annotation"
author: "Martin Morgan<br />
         Roswell Park Cancer Institute, Buffalo, NY<br />
         26-28 October, 2015"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{L3 -- Variant Calling and Annotation}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(VariantAnnotation)
    library(VariantFiltering)
    # library(VariantTools)
    # library(r5vc)
})
```

- Conceptual understanding of variant calling approaches and
  (primarily non-Bioconductor) tools. Exploratory approaches to
  variant calling in Bioconductor. Working with called variants, e.g.,
  classifying variants to genomic regions of interest and predicting
  effects of variants. Primary packages: [VariantTools][], [h5vc][],
  [VariantAnnotation][], [VariantFiltering][].


# Presentation: Variant Calling

## Overall work flow

Orientation

- Whole-genome versus exome-capture sequencing
- Germ-line versus somatic (cancer) variants
- DNA- versus RNA-seq

Key variant calling tools (non-_Bioconductor_)

- [BWA][] / [STAR][] / GMAP aligners
- [Picard][] FASTQ and BAM file management
- [GATK][] Variant calling, especially germline
- [muTect][] Somatic variant calling (challenge: low frequency variants)

[BWA]: http://bio-bwa.sourceforge.net/BWA-MEM
[STAR]: https://code.google.com/p/rna-star/STAR
[Picard]: https://broadinstitute.github.io/picard/index.html
[GATK]: https://www.broadinstitute.org/gatk
[muTect]: https://www.broadinstitute.org/cancer/cga/mutect

[GATK best practices][]

- Pre-processing
    - Align. High quality aligner, e.g., BWA MEM for DNA, STAR for
      RNASeq
    - Mark duplicates -- reads with identical start and end
      coordinates; likely from the same DNA template
    - Indel realignment -- aligners often 'solve' indel by introducing
      mismatches; realign by local consensus of reads. Less important
      for newer tools
    - Base recalibration -- machine learning approach to addressing
      systematic biases in base quality scores produced by sequencers,
      e.g., sequencer makes single-nucleotide quality assessments, but
      empirical data shows that dinucleotide biases
      exist. Lane-specific corrections.
- Variant discovery
    - sensitivity (minimize false negatives) vs. specifity (minimize
      false positives, esp. failing to reject artifacts). Two steps...
    - Call variants (maximize sensitivity).
          - Per-sample, followed by joint genotyping across samples.
          - Current state-of-the-art: haplotype _assembly_ in active
            regions (germ-line only)
    - Filter called variants (tune specificity).
          - Ideal: variant quality score recalibration -- apply
            machine learning to quality of called variants. Requires
            large callsets (exomes / genomes) and highly curateed
            known variants.
          - In practice: _ad hoc_ filter thresholds
- Callset refinement
    - Annotation
          - especially _functional_, e.g., location relative to coding
            sequence
    - Evaluation

## _Bioconductor_ opportunities

- [h5vc][] for managing nucleotide-level summaries
- [VariantTools][] for custom calling, filtering and annotation; e.g., 
  [VariantTools BioC2014 Tutorial](http://bioconductor.org/help/course-materials/2014/BioC2014/Lawrence_Tutorial.pdf)
- Tools for analysis of called variants (below)

# Called Variants

Classes -- GenomicRanges-like behavior

- _VCF_ -- 'wide'
- _VRanges_ -- 'tall'

Functions and methods

- I/O and filtering: `readVcf()`, `readGeno()`, `readInfo()`,
  `readGT()`, `writeVcf()`, `filterVcf()`
- Annotation: `locateVariants()` (variants overlapping ranges),
  `predictCoding()`, `summarizeVariants()`
- SNPs: `genotypeToSnpMatrix()`, `snpSummary()`

Related packages

- [ensemblVEP][] 
    - Forward variants to Ensembl Variant Effect Predictor
- [VariantTools][], [h5vc][]
    - Call variants
- [VariantFiltering][]
    - Filter variants using criteria such as coding consequence, MAF,
       ..., inheritance model

Reference

- Obenchain, V, Lawrence, M, Carey, V, Gogarten, S, Shannon, P, and
  Morgan, M. VariantAnnotation: a Bioconductor package for exploration
  and annotation of genetic variants. Bioinformatics, first published
  online March 28, 2014
  [doi:10.1093/bioinformatics/btu168](http://bioinformatics.oxfordjournals.org/content/early/2014/04/21/bioinformatics.btu168)


## Interactivities: exploring VCF files with _VariantAnnotation_

### What is a VCF file?

Header

    ##fileformat=VCFv4.2
    ##fileDate=20090805
    ##source=myImputationProgramV3.1
    ##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta
    ##contig=<ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species="Homo sapiens",taxonomy=x>
    ##phasing=partial
    ##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
    ##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">
    ...
    ##FILTER=<ID=q10,Description="Quality below 10">
    ##FILTER=<ID=s50,Description="Less than 50% of samples have data">
    ...
    ##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
    ##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
        
Location

    #CHROM POS     ID        REF    ALT     QUAL FILTER ...
    20     14370   rs6054257 G      A       29   PASS   ...
    20     17330   .         T      A       3    q10    ...
    20     1110696 rs6040355 A      G,T     67   PASS   ...
    20     1230237 .         T      .       47   PASS   ...
    20     1234567 microsat1 GTC    G,GTCT  50   PASS   ...
        
Variant INFO

    #CHROM POS     ...	INFO                              ...
    20     14370   ...	NS=3;DP=14;AF=0.5;DB;H2           ...
    20     17330   ...	NS=3;DP=11;AF=0.017               ...
    20     1110696 ...	NS=2;DP=10;AF=0.333,0.667;AA=T;DB ...
    20     1230237 ...	NS=3;DP=13;AA=T                   ...
    20     1234567 ...	NS=3;DP=9;AA=G                    ...
  
Genotype FORMAT and samples

    ... POS     ...  FORMAT      NA00001        NA00002        NA00003
    ... 14370   ...  GT:GQ:DP:HQ 0|0:48:1:51,51 1|0:48:8:51,51 1/1:43:5:.,.
    ... 17330   ...  GT:GQ:DP:HQ 0|0:49:3:58,50 0|1:3:5:65,3   0/0:41:3
    ... 1110696 ...  GT:GQ:DP:HQ 1|2:21:6:23,27 2|1:2:0:18,2   2/2:35:4
    ... 1230237 ...  GT:GQ:DP:HQ 0|0:54:7:56,60 0|0:48:4:51,51 0/0:61:2
    ... 1234567 ...  GT:GQ:DP    0/1:35:4       0/2:17:2       1/1:40:3

### Input and manipulation

Input and manipulation: [VariantAnnotation][]

```{r}
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
```

Headers

- `scanVcfHeader()`

    ```{r}
    scanVcfHeader(fl)
    info(scanVcfHeader(fl))
    ```

Data

- `readVcf()`
- Accessors, e.g., `info()`, `geno()`

    ```{r}
    vcf <- readVcf(fl, "hg19")
    vcf
    head(info(vcf))
    geno(vcf)
    head(geno(vcf)[["GT"]])
    ```

Restrictions

- `readInfo()`: simple vector of 
- `readGeno()`: simple matrix of genotypes

    ```{r}
    head(readGeno(fl, "GT"))
    ```

- `ScanVcfParam()`: limit input to specific `fixed`, `info`, `geno`,
  fields, `samples`, and genomic coordinates (`which`)

Chunks

- Read in chunks to limit memory consumption while allowing relatively
  efficient 'vectorized' computation.
  
    ```{r}
    dim(readVcf(fl, "hg19"))     # 10376 records

    vcf <- open(VcfFile(fl, yieldSize=10000))
    dim(readVcf(vcf, "hg19"))    # 10000 records...
    dim(readVcf(vcf, "hg19"))    # 376 records...
    dim(readVcf(vcf, "hg19"))    # 0 records, all done
    ```

### Annotation

Built-in

- Read variants from a VCF file, and annotate with respect to a known
  gene model
  
    ```{r vcf, message=FALSE}
    ## input variants
    library(VariantAnnotation)
    fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
    vcf <- readVcf(fl, "hg19")
    seqlevels(vcf) <- "chr22"
    ## known gene model
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    coding <- locateVariants(rowRanges(vcf),
        TxDb.Hsapiens.UCSC.hg19.knownGene,
        CodingVariants())
    head(coding)
    ```

Via additional resources

- [ensemblVEP][]
- SIFT, PolyPhen
- [myvariant][]

## Lab: _VariantFiltering_

For this lab, please work through the excellent [vignette] in the
[VariantFiltering][] package.

[VariantAnnotation]: https://bioconductor.org/packages/VariantFiltering
[VariantFiltering]: https://bioconductor.org/packages/VariantFiltering
[VariantTools]: https://bioconductor.org/packages/VariantTools
[ensemblVEP]: https://bioconductor.org/packages/ensemblVEP
[h5vc]: https://bioconductor.org/packages/h5vc
[myvariant]: https://bioconductor.org/packages/myvariant

[vignette]: http://bioconductor.org/packages/3.2/bioc/vignettes/VariantFiltering/inst/doc/usingVariantFiltering.pdf

[GATK best practices]: https://www.broadinstitute.org/gatk/guide/best-practices.php
