---
title: "L4 -- RNA-Seq Differential Expression"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{L4 -- RNA-Seq Differential Expression}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(DESeq2)
    library(edgeR)
})
```

Author: Martin Morgan
  (<a href="mailto:mtmorgan@fhcrc.org">mtmorgan@fhcrc.org</a>)<br />
Date: 26-28 October, 2015<br />

- Bioconductor known-gene RNA-seq differential expression work flow,
  from aligned reads to differential expression of genes. Important
  statistical issues and their resolution. Placing results of
  differential expression analysis into biological context. Brief
  discussion of novel-gene and transcript-level RNAseq differential
  expression analysis. Primary packages: [DESeq2][], [edgeR][].
