---
title: "L6 -- Visualization"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{L6 - Visualization}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(Gviz)
    library(ReportingTools)
    library(shiny)
    library(ggbio)
})
```

Author: Martin Morgan
  (<a href="mailto:mtmorgan@fhcrc.org">mtmorgan@fhcrc.org</a>)<br />
Date: 26-28 October, 2015<br />

- Challenges to visualizing genomic data. Opportunities for familiar
  and novel genomic visualizations in Bioconductor. Approaches to
  interactive visualization with shiny. Adopting strategies for
  reproducible and collaborative research. Primary packages: [Gviz][],
  [ReportingTools][], [shiny][], ([ggbio][]).


# Interactivity: stick figures

Create a fun shiny application

Advanced: implement _BAMSpector_ to explore BAM files in the context
of known gene annotation.

# Lab: _ReportingTools_

Transform the RNA-seq differential expression work flow into a
_ReportingTools_ report.