---
title: "L6 -- Visualization"
author: "Martin Morgan<br />
         Roswell Park Cancer Institute, Buffalo, NY<br />
         26-28 October, 2015"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{L6 - Visualization}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

- Challenges to visualizing genomic data. Opportunities for familiar
  and novel genomic visualizations in Bioconductor. Approaches to
  interactive visualization with shiny. Adopting strategies for
  reproducible and collaborative research. Primary packages: [Gviz][],
  [shiny][].

# Three approaches to visualizing data

```{r}
fl <- "BRFSS-subset.csv"
brfss <- read.csv(fl)
brfss2010 <- brfss[brfss$Year == "2010",]
```

## 'base' graphics

```{r}
## set layout, capture old options
opar <- par(mfcol=c(1, 2))
plot(sqrt(Weight) ~ Height, brfss2010[brfss2010$Sex == "Female", ],
    main="2010, Female")
plot(sqrt(Weight) ~ Height, brfss2010[brfss2010$Sex == "Male", ],
    main="2010, Male")
## restore old options
par(mfcol=c(1, 1))
```

## _lattice_

```{r}
library(lattice)
xyplot(sqrt(Weight) ~ Height | Sex, brfss2010)
bwplot(sqrt(Weight) ~ factor(Year) | Sex, brfss, panel=panel.violin)
```

## _ggplot2_

- Create a plot with `ggplot()`
- Add _layers_ (e.g., `geom_density()`) and _aesthetics_ (e.g., `aes()`)
- [ggplot2 documentation][]

```{r}
library(ggplot2)
ggplot(brfss2010) + geom_density(alpha=.2) + aes(sqrt(Weight), fill=Sex)
```

[ggplot2 documentation]: http://docs.ggplot2.org/current/

### Exercise

Revisit the DESeq2 workflow from yesterday

```{r}
library(DESeq2)
library(airway)
data(airway)
dds = DESeqDataSet(airway, design= ~ cell + dex)
dds = DESeq(dds)
res = results(dds)
```

The class of `res` is `DataFrame`, but to work with `ggplot2` we need
to coerce it to a `data.frame`.

```{r}
df <- as.data.frame(res)
```

Use `ggplot2` to create an MA plot, with log fold change on the y-axis
plotted against log base mean on the x-axis.

```{r, eval=FALSE}
ggplot(???, aes(???, ???)) + geom_point()
```

One problem when plotting many points is that points overlap. Modify
the plot by using the `stat_binhex()` transformation. Explore other
approaches to effectively displaying many points (e.g.,
`stat_density2d()`; the `alpha` argument to `geom_point()`)?

Check out the [HilbertCurve][] package for addressing another common
problem. As an illustration, here we calculate GC content in 1000nt
'tiles' across chromosome 1...

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
len = seqlengths(BSgenome.Hsapiens.UCSC.hg19)["chr1"]
tiles = tileGenome(len, tilewidth=1000, cut.last.tile.in.chrom=TRUE)
seq = getSeq(Hsapiens, tiles)
gc = letterFrequency(seq, "GC")
maxgc = max(gc)
```

...and then plot GC content as a Hilbert curve.

```{r, message=FALSE}
library(HilbertCurve)
library(circlize)
library(RColorBrewer)

col_fun=colorRamp2(seq(0, maxgc, length.out=9), brewer.pal(9, "YlOrRd"))
hc = HilbertCurve(1, len, 6, mode="pixel")
hc_layer(hc, tiles, col=col_fun(gc))
```

[HilbertCurve]: https://bioconductor.org/packages/HilbertCurve

# Genomic visualizations

## _Gviz_

### Exercise: Review section 2 of the [Gviz][] vignette

### Exercise: Gene model and BAM data

Visualize a BAM file from [RNAseqData.HNRNPC.bam.chr14][] around the
HNRNPC gene. Do this by creating an gene region track...

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
genes <- GeneRegionTrack(TxDb.Hsapiens.UCSC.hg19.knownGene, chromosome="chr14")
```
   
...a data track...

```{r}
library(RNAseqData.HNRNPC.bam.chr14)
fl <- RNAseqData.HNRNPC.bam.chr14_BAMFILES[1]
data <- DataTrack(range=fl, name=names(fl),
    type="histogram", window= -1,
    genome="hg19", chromosome="chr14")
```
 
...the region of interest for HNRNPC...

```{r}
library(org.Hs.eg.db)
egid <- mapIds(org.Hs.eg.db, "HNRNPC", "ENTREZID", "SYMBOL")
gene <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)[egid]
```

and plotting
 
```{r}
plotTracks(c(genes, data),
    from=start(gene), to=end(gene), chr="chr14", windowSize=30)
```

Add additional BAM files to the figure. Can you adjust the gene being
displayed?

# Interactivity

Create a fun shiny application

Advanced: implement _BAMSpector_ to explore BAM files in the context
of known gene annotation.
